docs/architecture.md

```diff
@@ -0,0 +1,241 @@
+# Loki's Architecture
+
+This document will expand on the information detailed in the [Loki
+Overview](overview/README.md).
+
+## Multi Tenancy
+
+All data - both in memory and in long-term storage - is partitioned by a
+tenant ID, pulled from the `X-Scope-OrgID` HTTP header in the request when Loki
+is running in multi-tenant mode. When Loki is **not** in multi-tenant mode, the
+header is ignored and the tenant ID is set to "fake", which will appear in the
+index and in stored chunks.
+
+## Modes of Operation
+
+![modes_diagram](modes_of_operation.png)
+
+Loki has a set of components (defined below in [Components](#components)) which
+are internally referred to as modules. Each component spawns a gRPC server for
+internal traffic and an HTTP/1 server for external API requests. All components
+come with an HTTP/1 server, but most only expose readiness, health, and metrics
+endpoints.
+
+While each component can be run in a separate process, Loki also supports running
+all components in a single process. Running all components in a single process is
+referred to as "single process" mode (or sometimes "single binary" or monolithic
+mode), while the other mode is the "horizontally" scalable, or microservices
+mode.
+
+When Loki runs in single process mode, individual components continue to
+communicate to one another over gRPC using the gRPC listen port of the overall
+process.
+
+Single process mode is ideally suited for local development, small workloads,
+and for evaluation purposes. Monolithic mode can be scaled with multiple
+processes with the following limitations:
+
+1. Local index and local storage cannot currently be used as they are not safe
+   for concurrent access by multiple processes.
+2. Individual components cannot be scaled independently, so it is not possible
+   to have more read components than write components.
+
+## Components
+
+### Distributor
+
+The **distributor** service is responsible for handling incoming streams by
+clients. It's the first stop in the write path for log data. Once the
+distributor receives a set of streams, each stream is validated for correctness
+and to ensure that it is within the configured tenant (or global) limits. Valid
+chunks are then split into batches and sent to multiple [ingesters](#ingester)
+in parallel.
+
+#### Hashing
+
+Distributors use consistent hashing in conjunction with a configurable
+replication factor to determine which instances of the ingester service should
+receive a given stream.
+
+A stream is a set of logs associated to a tenant and a unique labelset. The
+stream is hashed using both the tenant ID and the labelset and then the hash is
+used to find the ingesters to send the stream to.
+
+A hash ring stored in [Consul](https://www.consul.io) is used to achieve
+consistent hashing; all [ingesters](#ingester) register themselves into the hash
+ring with a set of tokens they own. Each token is a random unsigned 64-bit
```

Comment:
It's 32-bit (see ring.proto), but maybe it is too low-level detail for this document anyway.